dex/elrond_dex_farm/src/farm_token.rs

```diff
@@ -0,0 +1,197 @@
+                if token_id.is_egld() && returned_tokens > 0 {
+                    let _ = self.send().direct_egld(caller, &returned_tokens, &[]);
+                }
+            }
+        }
+    }
+
+    #[endpoint(setLocalRolesFarmToken)]
+    fn set_local_roles_farm_token(&self) -> SCResult<AsyncCall<Self::SendApi>> {
+        require!(self.is_active(), "Not active");
+        self.require_permissions()?;
+        require!(!self.farm_token_id().is_empty(), "No farm token issued");
+
+        let token = self.farm_token_id().get();
+        Ok(self.set_local_roles(token))
+    }
+
+    fn set_local_roles(&self, token: TokenIdentifier) -> AsyncCall<Self::SendApi> {
+        ESDTSystemSmartContractProxy::new_proxy_obj(self.send())
+            .set_special_roles(
+                &self.blockchain().get_sc_address(),
+                &token,
+                &[
+                    EsdtLocalRole::NftCreate,
+                    EsdtLocalRole::NftAddQuantity,
+                    EsdtLocalRole::NftBurn,
+                ],
+            )
+            .async_call()
+            .with_callback(self.callbacks().change_roles_callback())
+    }
+
+    #[callback]
+    fn change_roles_callback(&self, #[call_result] result: AsyncCallResult<()>) {
+        match result {
+            AsyncCallResult::Ok(()) => {
+                self.last_error_message().clear();
+            }
+            AsyncCallResult::Err(message) => {
+                self.last_error_message().set(&message.err_msg);
+            }
+        }
+    }
+
+    fn decode_attributes(
+        &self,
+        attributes_raw: &BoxedBytes,
+    ) -> SCResult<FarmTokenAttributes<Self::BigUint>> {
+        let attributes =
+            <FarmTokenAttributes<Self::BigUint>>::top_decode(attributes_raw.as_slice());
+        match attributes {
+            Result::Ok(decoded_obj) => Ok(decoded_obj),
+            Result::Err(_) => {
+                return sc_error!("Decoding error");
+            }
+        }
+    }
+
+    fn get_farm_attributes(
+        &self,
+        token_id: &TokenIdentifier,
+        token_nonce: u64,
+    ) -> SCResult<FarmTokenAttributes<Self::BigUint>> {
+        let token_info = self.blockchain().get_esdt_token_data(
+            &self.blockchain().get_sc_address(),
+            token_id,
+            token_nonce,
+        );
+
+        let farm_attributes = token_info.decode_attributes::<FarmTokenAttributes<Self::BigUint>>();
+        match farm_attributes {
+            Result::Ok(decoded_obj) => Ok(decoded_obj),
+            Result::Err(_) => {
+                return sc_error!("Decoding error");
+            }
+        }
+    }
+
+    fn create_farm_tokens(
+        &self,
+        farm_amount: &Self::BigUint,
+        farm_token_id: &TokenIdentifier,
+        attributes: &FarmTokenAttributes<Self::BigUint>,
+    ) -> Nonce {
+        self.nft_create_tokens(farm_token_id, farm_amount, attributes);
+        self.increase_nonce()
```

Comment:
we cannot keep on changing the contract code every time something gets fixed in Arwen

Reply:
I just said it is not needed right now. And yes. sometimes contracts can be simplified if there are improvements in Arwen.

Reply:
if we'll make the change here, we should make the same in all the contracts to keep consistency. i'd say it stays like this because it was audited and because of time issues